fix PR and refactor overall structure

igorlfs opened the pull request:
TODO

- [ ] ~Fix usage of `is_coloring_valid` in GRASP/PR~
- [x] Add docs
  - [x] assign_color()
  - [x] improve_phase()
- [x] Add tests
  - [x] count_forbidden_per_vertex()
  - [x] get_coloring_from_class_list()
  - [ ] ~assign_color()~
  - [x] improve_phase()

igorlfs committed:
fix PR and refactor overall structure

igorlfs commented on src/algorithms/grasp_pr.rs:
```diff
@@ -1,60 +1,38 @@
-            new_coloring[vertex] = best_coloring[vertex];
-
-            if count_forbidden_per_vertex(graph, &new_coloring, vertex) > 0 {
+            if !is_coloring_valid(graph, &new_coloring) {
                 continue;
             }
```
Isso aqui não deveria ser necessário. Esse `is_coloring_valid` chama o `is_valid_color_assignment` para todos os vértices. Só que como a gente só tá mudando um vértice de cada vez, só deveria ser necessário usar o `is_valid_color_assignment` no vértice que estamos alterando. Por algum motivo, isso não dá certo.

igorlfs replied:
Na verdade, conforme a gente troca a cor dos vértices, é possível "herdar" colorações inválidas. Por exemplo: se você chegou na coloração (inválida) 1--1--1 a partir de uma troca e, numa iteração seguinte, você troca o primeiro vértice obtendo 2--1--1, você ainda tem uma coloração inválida. É por isso que a gente sempre precisa chegar todos os vértices. Mas isso é custoso. Como a gente só troca a melhor solução quando o número de cores melhora, e isso é menos custoso de calcular, @vinisilvag, vc acha que seria uma boa ideia primeiro checar se a coloração melhora, gastando menos cores (maioria das vezes não vai) e depois checar se ela é valida?

vinisilvag replied:
Definitivamente, é uma ótima alternativa pra diminuir o custo computacional dessa parte. Por mim pode ser desse jeito que você sugeriu mesmo.

igorlfs committed:
refactor: cleaner docs and more tests

igorlfs committed:
test and docs pt2

igorlfs commented on src/algorithms/grasp_pr.rs:
```diff
@@ -1,60 +1,38 @@
-        let coloring = get_coloring_from_class_list(num_vertices, &solution.1);
-        let mut difference = simmetric_difference(&best_coloring, &coloring);
+    while let Some(solution) = solutions.pop() {
+        let original_best_coloring = best_solution.1.clone();
```
Outra coisa é que a gente está trocando a solução ótima dinamicamente, ou seja: se a gente troca a solução ótima na iteração `i`, na iteração `i+1` vamos usar essa nova solução ótima. Só que eu não sei se esse é o comportamento esperado do algoritmo, eu fiz isso isso por engano. Pra mim seria mais natural usar a solução ótima original ao longo de todo o algoritmo. O que você acha, @vinisilvag?

vinisilvag replied:
De fato parece mais natural testar a solução original sempre, mas não me parece que trocar a melhor solução assim tenha tanto impacto no resultado. Tipo, se uma solução intermediária melhora, me parece ok continuar testando ela contra as soluções restantes. Por mim, pelo menos a princípio, a gente podia deixar assim e se for o caso, mais pra frente, a gente testa a implementação que fixa a solução original contra algumas poucas instâncias só pra ver se dá alguma diferença notável.

igorlfs committed:
fix(grasp_pr): some oddities

- Fix the order of solutions: since Rust's BinaryHeap is a max-heap, if
  we pop, we take the largest element and we want the smallest instead.
  So transform into a vec, reverse and then pop
- Clarify that the behavior of swapping the best coloring is intended
  (as discussed in the PR)
- Prefer count_colors() earlier so we avoid having to check viability as
  often
